pkg/controller/component/lifecycle/kbagent.go

```diff
@@ -0,0 +1,235 @@
+	la := &memberJoin{
+		synthesizedComp: a.synthesizedComp,
+		pod:             a.pod,
+	}
+	return a.checkedCallAction(ctx, cli, a.lifecycleActions.MemberJoin, la, opts)
+}
+
+func (a *kbagent) MemberLeave(ctx context.Context, cli client.Reader, opts *Options) error {
+	la := &memberLeave{
+		synthesizedComp: a.synthesizedComp,
+		pod:             a.pod,
+	}
+	return a.checkedCallAction(ctx, cli, a.lifecycleActions.MemberLeave, la, opts)
+}
+
+func (a *kbagent) DataDump(ctx context.Context, cli client.Reader, opts *Options) error {
+	la := &dataDump{}
+	return a.checkedCallAction(ctx, cli, a.lifecycleActions.DataDump, la, opts)
+}
+
+func (a *kbagent) DataLoad(ctx context.Context, cli client.Reader, opts *Options) error {
+	la := &dataLoad{}
+	return a.checkedCallAction(ctx, cli, a.lifecycleActions.DataLoad, la, opts)
+}
+
+func (a *kbagent) AccountProvision(ctx context.Context, cli client.Reader, opts *Options, args ...any) error {
+	la := &accountProvision{args: args}
+	return a.checkedCallAction(ctx, cli, a.lifecycleActions.AccountProvision, la, opts)
+}
+
+func (a *kbagent) checkedCallAction(ctx context.Context, cli client.Reader,
+	handler *appsv1alpha1.LifecycleActionHandler, la lifecycleAction, opts *Options) error {
+	if handler == nil || handler.CustomHandler == nil {
+		return errors.Wrap(ErrActionNotDefined, la.name())
+	}
+	return a.callAction(ctx, cli, handler.CustomHandler, la, opts)
+}
+
+func (a *kbagent) callAction(ctx context.Context, cli client.Reader, spec *appsv1alpha1.Action, la lifecycleAction, opts *Options) error {
+	req, err1 := a.buildActionRequest(ctx, cli, la, opts)
+	if err1 != nil {
+		return err1
+	}
+	return a.callActionWithSelector(ctx, spec, la, req)
+}
+
+func (a *kbagent) buildActionRequest(ctx context.Context, cli client.Reader, la lifecycleAction, opts *Options) (*proto.ActionRequest, error) {
+	parameters, err := la.parameters(ctx, cli)
+	if err != nil {
+		return nil, err
+	}
+
+	req := &proto.ActionRequest{
+		Action:     la.name(),
+		Parameters: parameters,
+	}
+	if opts != nil {
+		if opts.NonBlocking != nil {
+			req.NonBlocking = opts.NonBlocking
+		}
+		if opts.TimeoutSeconds != nil {
+			req.TimeoutSeconds = opts.TimeoutSeconds
+		}
+		if opts.RetryPolicy != nil {
+			req.RetryPolicy = &proto.RetryPolicy{
+				MaxRetries:    opts.RetryPolicy.MaxRetries,
+				RetryInterval: opts.RetryPolicy.RetryInterval,
+			}
+		}
+	}
+	return req, nil
+}
+
+func (a *kbagent) callActionWithSelector(ctx context.Context, spec *appsv1alpha1.Action, la lifecycleAction, req *proto.ActionRequest) error {
+	pods, err := a.selectTargetPods(spec)
+	if err != nil {
+		return err
+	}
+	if len(pods) == 0 {
+		return fmt.Errorf("no available pod to call action %s", la.name())
+	}
+
+	// TODO: impl
+	//  - back-off to retry
+	//  - timeout
+	for _, pod := range a.pods {
+		cli, err1 := kbacli.NewClient(*pod)
+		if err1 != nil {
+			return err1
+		}
+		if cli == nil {
+			continue // not defined, for test only
```

Comment:
if the Pod doesn't have IP yet, cli would be nil. Should here return a error ?

Reply:
An error will be returned if the pod doesn't have IP assigned currently. This branch is added to cover the scenario that the pod has no kb-agent (port) defined, and this's impossible in production code, but there are too many test cases that mock pods to hack component and ITS status without a kb-agent container in it.